apollo-router/src/plugins/telemetry/mod.rs

```diff
@@ -1245,6 +1265,120 @@ impl Telemetry {
+                .filter_map(|value| decode_ftv1_trace(value.as_str()?))
+                .collect()
+        } else {
+            Vec::new()
+        }
+    }
+
+    // https://github.com/apollographql/apollo-server/blob/6ff88e87c52/packages/server/src/plugin/usageReporting/stats.ts#L283
+    fn per_type_stat(
+        traces: &[proto::reports::Trace],
+        field_level_instrumentation_ratio: f64,
+    ) -> HashMap<String, SingleTypeStat> {
+        fn recur(
+            per_type: &mut HashMap<String, SingleTypeStat>,
+            field_execution_weight: f64,
+            node: &proto::reports::trace::Node,
+        ) {
+            for child in &node.child {
+                recur(per_type, field_execution_weight, child)
+            }
+            let response_name = if let Some(ResponseName(response_name)) = &node.id {
+                response_name
+            } else {
+                return;
+            };
+            let field_name = if node.original_field_name.is_empty() {
+                response_name
+            } else {
+                &node.original_field_name
+            };
+            if field_name.is_empty()
+                || node.parent_type.is_empty()
+                || node.r#type.is_empty()
+                || node.start_time == 0
+                || node.end_time == 0
+            {
+                return;
+            }
+            let field_stat = per_type
+                .entry(node.parent_type.clone())
+                .or_default()
+                .per_field_stat
+                .entry(field_name.clone())
+                .or_insert_with(|| metrics::apollo::studio::SingleFieldStat {
+                    return_type: node.r#type.clone(), // not `Default::default()`’s empty string
+                    errors_count: 0,
+                    latency: Default::default(),
+                    estimated_execution_count: 0.0,
+                    requests_with_errors_count: 0,
+                });
+            let latency = Duration::from_nanos(node.end_time.saturating_sub(node.start_time));
+            field_stat
+                .latency
+                .increment_duration(Some(latency), field_execution_weight);
+            field_stat.estimated_execution_count += field_execution_weight;
+            field_stat.errors_count += node.error.len() as u64;
+            if !node.error.is_empty() {
+                field_stat.requests_with_errors_count += 1;
+            }
+        }
+
+        // For example, `field_level_instrumentation_ratio == 0.03` means we send a
+        // `apollo-federation-include-trace: ftv1` header with 3% of subgraph requests.
+        // To compensate, assume that each trace we recieve is representative of 33.3… requests.
+        // Metrics that recieve this treatment are kept as floating point values in memory,
+        // and converted to integers after aggregating values for a number of requests.
+        let field_execution_weight = 1.0 / field_level_instrumentation_ratio;
```

Comment:
I made a comment about this earlier today but maybe I did it not on the PR or something because I don't see it... anyway, just curious if this line can ever run with `field_level_instrumentation_ratio` zero.

Reply:
It’s on the commit, detached from the PR. I replied there: https://github.com/apollographql/router/commit/30a36c3837bba154e81537fe8d3c76d0f85464c5#r100085309

----

If `field_level_instrumentation_ratio` is zero:
* `rng.gen_bool(field_level_instrumentation_ratio)` will always return false
* `ApolloFtv1Handler` will always be disabled
* `store_ftv1` will always return early without reading `extensions.ftv1` from subgraph responses (which would likely not be there anyway since `request_ftv1` didn’t set a `apollo-federation-include-trace` header)
* This `per_type_stat` function would be called but with an empty `traces: &[proto::reports::Trace]`. So `field_execution_weight == f64::INFINITY` could happen but it won’t be added to any stats.

Reply:
This can indeed be zero, and we need to add code to account for this.

Reply:
Why not using `checked_div` ?

Reply:
`checked_div` is a method of integer types, where division by zero panics. Here `field_level_instrumentation_ratio` is a `f64`, where 1 / 0 is well defined to be positive infinity which is representable in f64. It would not be desirable to add it to actual stats, but as explained above this doesn’t happen.